File: Vector.java
/**
 * Created by sxh on 11/8/14.
 */
public class Vector{
    public static final double EPS = 0.01;

    public double[] myCoo = null;

    public Vector () {}

    public Vector (double[] coo) {
        myCoo = coo;
    }

    public Vector (Vector c) {
        myCoo = new double[c.myCoo.length];
        for (int i=0; i<myCoo.length; i++) myCoo[i] = c.myCoo[i];
    }

    public static Vector add (Vector a, Vector b) {
        if (a.myCoo.length != b.myCoo.length) throw new IllegalArgumentException();
        Vector result = new Vector();
        result.myCoo = new double[a.myCoo.length];
        for (int i=0; i< a.myCoo.length; i++) result.myCoo[i] = a.myCoo[i] + b.myCoo[i];
        return result;
    }

    public static Vector mul (Vector a, double c) {
        Vector result = new Vector();
        result.myCoo = new double[a.myCoo.length];
        for (int i=0; i< a.myCoo.length; i++) result.myCoo[i] = c * a.myCoo[i];
        return result;
    }

    public static Vector minus (Vector a) {return mul(a, -1);}

    public static Vector minus (Vector a, Vector b) {return add(a, minus(b));}

    public static double len2 (Vector a) {
        double result = 0.0;
        for (int i=0; i< a.myCoo.length; i++) result += a.myCoo[i]*a.myCoo[i];
        return result;
    }

    public static double len (Vector a) {return Math.sqrt(len2(a));}

    public static double prod (Vector a, Vector b) {
        if (a.myCoo.length != b.myCoo.length) throw new IllegalArgumentException();
        double result = 0.0;
        for (int i=0; i< a.myCoo.length; i++) result += a.myCoo[i]*b.myCoo[i];
        return result; /// (len(a) * len(b));
    }

    public static double angle (Vector a, Vector b) {
        return Vector.prod(a, b) / (Vector.len(a) * Vector.len(b));
    }

    public static boolean orthogonal(Vector a, Vector b) {
        return Math.abs(angle(a, b)) < EPS;
    }

    public Vector reflectWRT (Vector y) {return add(y, mul(this, - 2 * prod (this, y) / len2(this)));}

    @Override
    public boolean equals(Object o) {
        if (o == null) return false;
        if (o instanceof Vector) {
            Vector vo = (Vector) o;
            return (len(add(this, minus(vo))) < EPS);
        }
        throw new IllegalArgumentException();
    }

    @Override
    public int hashCode() {
        int result = 0;
        for (double aMyCoo : myCoo) {
            int k = (int) Math.round(aMyCoo / EPS);
            result = result * 31 + k;
        }
        return result;
    }

    @Override
    public String toString() {
        String result = "{";
        for (int i=0; i<myCoo.length; i++) {
            result += fmt(myCoo[i]);
            if (i < myCoo.length-1) result+=", ";
        }
        return result + "}";
    }

    public static String fmt(double d) {
        if(Math.abs((d - (long) d)) <= 0.00001) return String.format("%d",(long)d); else return String.format("%e",d);}
}

File: src/primitives/Vector.kt
class Vector constructor (val myCoo: DoubleArray) {
    constructor(len: Int): this(DoubleArray(len))
    constructor(vec: Vector): this(DoubleArray(vec.myCoo.size, {p: Int -> vec.myCoo[p]}))

    fun reflectWRT(y: Vector): Vector = add(y, mul(this, -2 * prod(this, y) / len2(this)))

    override fun equals(o: Any?): Boolean {
        if (o == null) return false
        if (o is Vector) return len(add(this, minus(o))) < EPS
        throw IllegalArgumentException()
    }

    override fun hashCode(): Int = myCoo.fold(0, {result: Int, p: Double -> result * 31 + Math.round(p / EPS).toInt()})

    override fun toString(): String {
        val result = StringBuilder("{")
        for (i in myCoo.indices) {
            result.append(fmt(myCoo[i]))
            if (i < myCoo.size - 1) result.append(", ")
        }
        return result.toString() + "}"
    }

    fun getReflection(): Matrix {
        val n = myCoo.size
        val m = Array(n) { DoubleArray(n) }

        var d = len2(this)
        if (Math.abs(d) < EPS) throw IllegalArgumentException("Can not reflect w.r.t. zero vector.")

        d = -2 / d

        for (i in 0 until n) { //column iterator
            val c = d * myCoo[i]
            for (j in 0 until n) { //row iterator
                m[j][i] = c * myCoo[j]
                if (i == j) m[j][i] += 1.0
            }
        }

        return Matrix(m)
    }


    companion object {
        val EPS = 0.01

        fun fmt(d: Double): String {
            return if (Math.abs(d - d.toLong()) <= 0.00001)
                String.format("%d", d.toLong())
            else
                String.format("%e", d)
        }


        fun add(a: Vector, b: Vector): Vector {
            if (a.myCoo.size != b.myCoo.size) throw IllegalArgumentException()
            val result = Vector(a.myCoo.size)
            for (i in a.myCoo.indices) result.myCoo[i] = a.myCoo[i] + b.myCoo[i]
            return result
        }

        fun mul(a: Vector, c: Double): Vector {
            val result = Vector(a.myCoo.size)
            for (i in a.myCoo.indices) result.myCoo[i] = c * a.myCoo[i]
            return result
        }

        fun minus(a: Vector): Vector = mul(a, -1.0)

        fun minus(a: Vector, b: Vector) = add(a, minus(b))

        fun len2(a: Vector): Double = a.myCoo.fold(0.0, {result: Double, b: Double -> result + b*b})

        fun len(a: Vector): Double = Math.sqrt(len2(a))

        fun prod(a: Vector, b: Vector): Double {
            if (a.myCoo.size != b.myCoo.size) throw IllegalArgumentException()
            return a.myCoo.zip(b.myCoo).fold(0.0, {result: Double, p: Pair<Double, Double> -> result + p.first * p.second }) 
        }

        fun angle(a: Vector, b: Vector): Double = prod(a, b) / (len(a) * len(b))

        fun orthogonal(a: Vector, b: Vector): Boolean = Math.abs(angle(a, b)) < EPS
    }
}